// discordx/build/cjs/decorators/classes/DComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DComponent = void 0;
const Method_js_1 = require("./Method.js");
/**
 * @category Decorator
 */
class DComponent extends Method_js_1.Method {
    _type;
    _id;
    _guilds;
    _botIds;
    get type() {
        return this._type;
    }
    get botIds() {
        return this._botIds;
    }
    set botIds(value) {
        this._botIds = value;
    }
    get id() {
        return this._id;
    }
    set id(value) {
        this._id = value;
    }
    get guilds() {
        return this._guilds;
    }
    set guilds(value) {
        this._guilds = value;
    }
    constructor(type, id, guilds, botIds) {
        super();
        this._type = type;
        this._id = id;
        this._guilds = guilds ?? [];
        this._botIds = botIds ?? [];
    }
    static create(type, id, guilds, botIds) {
        return new DComponent(type, id, guilds, botIds);
    }
    isId(text) {
        return typeof this.id === "string" ? this.id === text : this.id.test(text);
    }
    parseParams() {
        return [];
    }
}
exports.DComponent = DComponent;
