// discordx/build/esm/decorators/classes/DApplicationCommand.js
import { DApplicationCommandOption, resolveIPermissions } from "../../index.js";
import { Method } from "./Method.js";
/**
 * @category Decorator
 */
export class DApplicationCommand extends Method {
    _name;
    _description;
    _type;
    _defaultPermission;
    _options = [];
    _permissions = [];
    _guilds;
    _group;
    _subgroup;
    _botIds;
    get type() {
        return this._type;
    }
    set type(value) {
        this._type = value;
    }
    get botIds() {
        return this._botIds;
    }
    set botIds(value) {
        this._botIds = value;
    }
    get group() {
        return this._group;
    }
    set group(value) {
        this._group = value;
    }
    get subgroup() {
        return this._subgroup;
    }
    set subgroup(value) {
        this._subgroup = value;
    }
    get permissions() {
        return this._permissions;
    }
    set permissions(value) {
        this._permissions = value;
    }
    get guilds() {
        return this._guilds;
    }
    set guilds(value) {
        this._guilds = value;
    }
    get defaultPermission() {
        return this._defaultPermission;
    }
    set defaultPermission(value) {
        this._defaultPermission = value;
    }
    get name() {
        return this._name;
    }
    set name(value) {
        this._name = value;
    }
    get description() {
        return this._description;
    }
    set description(value) {
        this._description = value;
    }
    get options() {
        return this._options;
    }
    set options(value) {
        this._options = value;
    }
    constructor(name, type, description, defaultPermission, guilds, botIds) {
        super();
        this._name = name;
        this._type = type;
        this._description = description ?? this.name;
        this._defaultPermission = defaultPermission ?? true;
        this._guilds = guilds ?? [];
        this._botIds = botIds ?? [];
    }
    static create(name, type, description, defaultPermission, guilds, botIds) {
        return new DApplicationCommand(name, type, description, defaultPermission, guilds, botIds);
    }
    resolvePermissions(guild, command) {
        return resolveIPermissions(guild, command, this.permissions);
    }
    toSubCommand() {
        const option = DApplicationCommandOption.create(this.name, undefined, undefined, this.description, undefined, undefined, undefined, undefined, "SUB_COMMAND").decorate(this.classRef, this.key, this.method, this.from, this.index);
        option.options = this.options;
        return option;
    }
    async toJSON(command) {
        if (this.type !== "CHAT_INPUT") {
            const data = {
                defaultPermission: typeof this.defaultPermission === "boolean"
                    ? this.defaultPermission
                    : await this.defaultPermission.resolver(command),
                description: "",
                name: this.name,
                options: [],
                type: this.type,
            };
            return data;
        }
        const options = [...this.options]
            .reverse()
            .sort((a, b) => {
            if ((a.type === "SUB_COMMAND" || a.type === "SUB_COMMAND_GROUP") &&
                (b.type === "SUB_COMMAND" || b.type === "SUB_COMMAND_GROUP")) {
                return a.name < b.name ? -1 : 1;
            }
            return 0;
        })
            .map((option) => option.toJSON());
        const data = {
            defaultPermission: typeof this.defaultPermission === "boolean"
                ? this.defaultPermission
                : await this.defaultPermission.resolver(command),
            description: this.description,
            name: this.name,
            options: options,
            type: this.type,
        };
        return data;
    }
    parseParams(interaction) {
        return [...this.options].reverse().map((op) => {
            switch (op.type) {
                case "STRING":
                    return interaction.options.getString(op.name) ?? undefined;
                case "BOOLEAN":
                    return interaction.options.getBoolean(op.name) ?? undefined;
                case "NUMBER":
                    return interaction.options.getNumber(op.name) ?? undefined;
                case "INTEGER":
                    return interaction.options.getInteger(op.name) ?? undefined;
                case "ROLE":
                    return interaction.options.getRole(op.name) ?? undefined;
                case "CHANNEL":
                    return interaction.options.getChannel(op.name) ?? undefined;
                case "MENTIONABLE":
                    return interaction.options.getMentionable(op.name) ?? undefined;
                case "USER":
                    return (interaction.options.getMember(op.name) ??
                        interaction.options.getUser(op.name) ??
                        undefined);
                default:
                    return interaction.options.getString(op.name) ?? undefined;
            }
        });
    }
}
