// discordx/build/esm/decorators/classes/DDiscord.js
import { DIService } from "@discordx/di";
import { Decorator } from "@discordx/internal";
/**
 * @category Decorator
 */
export class DDiscord extends Decorator {
    _name;
    _description;
    _guards = [];
    _buttonComponents = [];
    _selectMenuComponents = [];
    _applicationCommands = [];
    _simpleCommands = [];
    _events = [];
    _defaultPermission = true;
    _permissions = [];
    _guilds = [];
    _botIds = [];
    get permissions() {
        return this._permissions;
    }
    set permissions(value) {
        this._permissions = value;
    }
    get botIds() {
        return this._botIds;
    }
    set botIds(value) {
        this._botIds = value;
    }
    get guilds() {
        return this._guilds;
    }
    set guilds(value) {
        this._guilds = value;
    }
    get defaultPermission() {
        return this._defaultPermission;
    }
    set defaultPermission(value) {
        this._defaultPermission = value;
    }
    get description() {
        return this._description;
    }
    set description(value) {
        this._description = value;
    }
    get name() {
        return this._name;
    }
    set name(value) {
        this._name = value;
    }
    get guards() {
        return this._guards;
    }
    set guards(value) {
        this._guards = value;
    }
    get applicationCommands() {
        return this._applicationCommands;
    }
    set applicationCommands(value) {
        this._applicationCommands = value;
    }
    get simpleCommands() {
        return this._simpleCommands;
    }
    set simpleCommands(value) {
        this._simpleCommands = value;
    }
    get buttons() {
        return this._buttonComponents;
    }
    set buttons(value) {
        this._buttonComponents = value;
    }
    get selectMenus() {
        return this._selectMenuComponents;
    }
    set selectMenus(value) {
        this._selectMenuComponents = value;
    }
    get events() {
        return this._events;
    }
    set events(value) {
        this._events = value;
    }
    get instance() {
        return DIService.instance.getService(this.from);
    }
    constructor(name, description) {
        super();
        this._name = name;
        this._description = description ?? name;
    }
    static create(name, description) {
        return new DDiscord(name, description);
    }
}
