// discordx/build/cjs/classes/SimpleCommandMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleCommandMessage = void 0;
const tslib_1 = require("tslib");
const crypto_1 = (0, tslib_1.__importDefault)(require("crypto"));
const discord_js_1 = require("discord.js");
const index_js_1 = require("../index.js");
/**
 * Simple command message
 */
class SimpleCommandMessage {
    prefix;
    argString;
    message;
    info;
    splitter;
    options = [];
    constructor(prefix, argString, message, info, splitter) {
        this.prefix = prefix;
        this.argString = argString;
        this.message = message;
        this.info = info;
        this.splitter = splitter;
        // empty constructor
    }
    get name() {
        return this.info.name;
    }
    get description() {
        return this.info.description;
    }
    /**
     * Resolve options
     */
    resolveOptions() {
        return this.info.parseParamsEx(this);
    }
    /**
     * Verify that all options are valid
     *
     * @returns
     */
    isValid() {
        return !this.options.includes(undefined);
    }
    /**
     * Get related commands
     *
     * @returns
     */
    getRelatedCommands() {
        const commandName = this.info.name.split(" ")[0];
        if (!commandName) {
            return [];
        }
        return index_js_1.MetadataStorage.instance.simpleCommands.filter((cmd) => cmd.name.startsWith(commandName) && cmd.name !== this.info.name);
    }
    /**
     * Send usage syntax for command
     *
     * @returns
     */
    sendUsageSyntax() {
        const maxLength = !this.info.options.length
            ? 0
            : this.info.options.reduce((a, b) => a.name.length > b.name.length ? a : b).name.length;
        const embed = new discord_js_1.MessageEmbed();
        embed.setColor(crypto_1.default.randomInt(654321));
        embed.setTitle("Command Info");
        embed.addField("Name", this.info.name);
        embed.addField("Description", this.info.description);
        // add aliases
        if (this.info.aliases.length) {
            embed.addField("Aliases", this.info.aliases.join(", "));
        }
        // add syntax usage
        embed.addField("Command Usage", "```" +
            this.prefix +
            this.name +
            ` ${this.info.options
                .map((op) => `{${op.name}: ${index_js_1.SimpleCommandOptionType[op.type]}}`)
                .join(" ")}` +
            "```");
        // add options if available
        if (this.info.options.length) {
            embed.addField("Options", "```" +
                this.info.options
                    .map((op) => `${op.name.padEnd(maxLength + 2)}: ${op.description}`)
                    .join("\n") +
                "```");
        }
        return this.message.reply({ embeds: [embed] });
    }
}
exports.SimpleCommandMessage = SimpleCommandMessage;
